### PR TITLE
Use ConnectWithRetries in PingPort.

### PR DESCRIPTION
**Pull request by dHannasch**
This is built atop https://github.com/ray-project/ray/pull/12976. Look there first.

## Why are these changes needed?

PingPort was originally hacked together as a way of checking an address without crashing if that address was inaccessible.
https://github.com/ray-project/ray/pull/12976 makes the normal connection process use a refactored function that tries to connect without crashing if that address is inaccessible.
With https://github.com/ray-project/ray/pull/12976 in play, PingPort can now use the exact same code that would actually be used to connect.
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by dHannasch** [comment]
The failing tests look to be unrelated to this change.

File: src/ray/gcs/redis_context.cc [hunk 1]
```diff
@@ -345,10 +346,22 @@ Status ConnectWithRetries(const std::string &address, int port,
   return Status::OK();
 }
 
+template <typename RedisContext, typename RedisConnectFunction>
+Status ConnectOrFail(const std::string &address, int port,
+                     const RedisConnectFunction &connect_function,
+                     RedisContext **context) {
+  std::string errorMessage = "";
+  Status status = ConnectWithRetries(address, port, connect_function, context);
+  if (!status.ok()) {
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/redis_context.cc:355]: Tiny: check this status with `Status::RedisError`

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -357,7 +370,7 @@ Status RedisContext::Connect(const std::string &address, int port, bool sharding
   RAY_CHECK(!redis_async_context_);
   RAY_CHECK(!async_redis_subscribe_context_);
 
-  RAY_CHECK_OK(ConnectWithRetries(address, port, redisConnect, &context_));
+  RAY_CHECK_OK(ConnectOrFail(address, port, redisConnect, &context_));
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/redis_context.cc:373]: `ConnectOrFail ` always returns OK, right?
can it return other Status while connect failed?

File: src/ray/gcs/redis_context.cc [hunk 3]
```diff
@@ -327,10 +327,11 @@ Status ConnectWithRetries(const std::string &address, int port,
       ConnectWithoutRetries(address, port, connect_function, context, errorMessage);
   while (!status.ok()) {
     if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
-      RAY_LOG(FATAL) << RayConfig::instance().redis_db_connect_retries() << " attempts "
-                     << "to connect have all failed. The last error message was: "
-                     << errorMessage;
-      break;
+      std::string fatalError = "";
```
Comment (WangTaoTheTonic) [inline src/ray/gcs/redis_context.cc:330]: Tiny: rename to `error_msg`? We don't use camel naming with variables in Ray.

File: src/ray/gcs/redis_context.cc (hunk 2 above)
Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: `ConnectOrFail` is intended to be a drop-in replacement for the old `ConnectWithRetries`, which indeed never actually returned anything other than OK.

File: src/ray/gcs/redis_context.cc (hunk 1 above)
Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:355]: > check this status with `Status::RedisError`

Sorry, what? I don't understand.

...Do you mean `IsRedisError()`, perhaps? If that's what you mean, shouldn't it fail no matter what kind of error? (We might later want to return different kinds of errors.)

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:355]: Yeah. For `RedisError` we should fail but for other tpyes of error we might do other actions(now we only have one error so that's an assumption).
For now it looks like we should fail whatever error happens, but i think it's a good practice to tell difference between types of error, for future changes.

File: src/ray/gcs/redis_context.cc (hunk 2 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: `ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status. The naming is kind of contrarily, i think.

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: > `ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status. The naming is kind of contrarily, i think.

Uh...you think so? The point of renaming `ConnectWithRetries` to `ConnectOrFail` and giving the name `ConnectWithRetries` to a sub-function was that it seemed kind of unexpected that a function called "ConnectWithRetries" might crash the entire program. The point of the name "ConnectOrFail" is to get across that when you call this function, it might exit.

Is it the word "Fail" that isn't working for you? Maybe "ConnectOrAbort" or "ConnectOrFatal" or something?

> `ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status.

The set of situations where `ConnectOrFail` will return OK is literally exactly the same set of situations where `ConnectWithRetries` returns OK. The only difference in their behavior is what happens if they *don't* return OK.

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: I think from point of naming, methods suffixed with `Fail` should fail(returns status that is not ok) if the condition was not satisfied, while that suffixed with `Retries` should retry for couple of times if the condition was not satisfied for temporarily.

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: Okay, but what we have is a function that exits the program with `RAY_LOG(FATAL)`. I agree that we need clarity between the version that retries and the version that doesn't. We do indeed have ConnectWithRetries versus ConnectWithoutRetries to clarify whether a function retries or not, and it sounds like that is something you want, right? So the question remaining is what to call the function that might exit the program. You're talking about *what* a function returns, but what's special about this function is that it might never return at all. I'm perfectly happy to call it anything, just as long as there is some indication somewhere in there that it might not return. ConnectOrAbort, ConnectOrFatal, ConnectOrExit, anything.

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: Thanks for your try in the tiny naming thing :)
for b658a7f i think we just keep the if branch, like 
`if (status.IsRedisError()) {
....
}`, while the `else` branch is not necessary.

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: > i think we just keep the if branch, like
> `if (status.IsRedisError()) { .... }`, while the `else` branch is not necessary.

Uh, I mean, we can, and that'll be harmless at the moment, but I thought the point was to future-proof against adding other possible errors? The function currently called `ConnectOrFail`  is the heir to the original `ConnectWithRetries` which could never return without connecting successfully. If we change that, then it will be possible for the function to return *without* connecting. I assume that the reason the function was written that way in the first place was because all the code after it relies on the fact that the connection was definitely successful.

It seems like the *safest* thing to do when the function encounters an unanticipated type of error is to halt and catch fire. Allowing the code to just keep going and ignore the failed connection would mean we could get unexpected behavior in the code following the connection call.

Basically, what we started with is:
```
Connect()
assert IsConnected()
(lots more code)
```
It sounds like you're talking about
```
Connect()
assert not IsOneParticularKindOfError()
(lots more code)
```

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: For now `ConnectWithRetries` only returns OK or RedisError, i think in the codes where it's invoked it's ok we only need to check these two status. In the future if more status returned, we can add their handle logic ( which is perhaps a different logic).

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: Like in https://github.com/ray-project/ray/pull/12976?

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: https://github.com/ray-project/ray/pull/12976/files#diff-478a807e3003772df7d270ff7ba820665462d5c82d008dff07dc0f3d4747af92R355
Why just write as like:
```
  if (status.IsRedisError()) {
      RAY_LOG(FATAL) << status.message();
  }
```

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: I think that silently swallowing all other possible errors and returning OK without even a hint that something is happening there is asking for trouble, but I don't feel very strongly about it.
https://github.com/ray-project/ray/pull/12976/commits/66b2bf85a93e6050e5134cfcae708382fd73a3fc

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: Ok. How about this?
```
  if (!status.ok()) {
      RAY_LOG(FATAL) << "Error when connect to redis with status " << status <<  " and error message "   status.message();
  }
```
You can modify the sentence as you want, with printing error code and message :)

Comment (dHannasch) [reply src/ray/gcs/redis_context.cc:373]: Oh! I think I completely misunderstood what you were driving at. Are you just concerned that some future `.message()` might not contain enough information? I can use `ToString()`, that should include all the relevant information from the status, right? Does https://github.com/ray-project/ray/pull/12976 look good?

Comment (WangTaoTheTonic) [reply src/ray/gcs/redis_context.cc:373]: LGTM

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```
